File: pydardcor/engine/lsp_client.py
"""Phase C: Language Server Protocol client for Dardcor Code."""

import os
import json
import subprocess
import threading
from typing import Optional, Dict, Any, List, Callable


class LSPClient:
    def __init__(self, language_id: str, server_cmd: List[str], workspace_path: str = ""):
        self.language_id = language_id
        self._cmd = server_cmd
        self._workspace_path = workspace_path
        self._process: Optional[subprocess.Popen] = None
        self._request_seq = 0
        self._pending: Dict[int, Any] = {}
        self._lock = threading.Lock()
        self._initialized = False
        self._diagnostics_handler: Optional[Callable] = None
        self._completion_handler: Optional[Callable] = None
        self._hover_handler: Optional[Callable] = None

    def start(self) -> bool:
        if self._process and self._process.poll() is None:
            return True

        try:
            self._process = subprocess.Popen(
                self._cmd,
                stdin=subprocess.PIPE,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True,
                encoding="utf-8",
                errors="replace",
                bufsize=1,
                cwd=self._workspace_path or None,
            )
            reader = threading.Thread(target=self._read_loop, daemon=True)
            reader.start()
            self._initialize()
            return True
        except Exception:
            return False

    def _initialize(self):
        params = {
            "processId": os.getpid(),
            "rootUri": f"file:///{self._workspace_path}" if self._workspace_path else None,
            "capabilities": {
                "textDocument": {
                    "synchronization": {"didOpen": True, "didChange": True, "didSave": True, "willSave": False},
                    "completion": {"completionItem": {"snippetSupport": True}},
                    "hover": {"contentFormat": ["markdown", "plaintext"]},
                    "definition": {"dynamicRegistration": False},
                    "references": {"dynamicRegistration": False},
                    "publishDiagnostics": {"relatedInformation": True},
                },
                "workspace": {"workspaceFolders": True},
            },
            "workspaceFolders": [{"uri": f"file:///{self._workspace_path}", "name": os.path.basename(self._workspace_path)}] if self._workspace_path else [],
        }
        resp = self._send_request_sync("initialize", params)
        if resp is not None:
            self._initialized = True
            self._send_notification("initialized", {})

    def _read_loop(self):
        try:
            content_length = None
            buffer = ""
            while True:
                line = self._process.stdout.readline()
                if not line:
                    break
                line = line.strip()
                if line.startswith("Content-Length:"):
                    content_length = int(line.split(":")[1].strip())
                elif line == "" and content_length is not None:
                    body = self._process.stdout.read(content_length)
                    content_length = None
                    if body:
                        msg = json.loads(body)
                        self._handle_message(msg)
        except Exception:
            pass

    def _handle_message(self, msg: dict):
        if "id" in msg and "method" not in msg:
            with self._lock:
                if msg["id"] in self._pending:
                    self._pending[msg["id"]].set_result(msg.get("result"))
        elif "method" in msg:
            if msg["method"] == "textDocument/publishDiagnostics":
                uri = msg.get("params", {}).get("uri", "")
                diagnostics = msg.get("params", {}).get("diagnostics", [])
                if self._diagnostics_handler:
                    self._diagnostics_handler(uri, diagnostics)

    def _send_request_sync(self, method: str, params: Any = None, timeout: float = 10.0) -> Any:
        if not self._process or self._process.poll() is not None:
            return None

        self._request_seq += 1
        seq = self._request_seq

        msg = {"jsonrpc": "2.0", "id": seq, "method": method, "params": params or {}}
        body = json.dumps(msg)
        header = f"Content-Length: {len(body.encode('utf-8'))}\r\n\r\n"

        try:
            self._process.stdin.write(header + body)
            self._process.stdin.flush()
        except Exception:
            return None

        result_holder = [None]
        event = threading.Event()

        class Future:
            def set_result(self, r):
                result_holder[0] = r
                event.set()

        future = Future()
        with self._lock:
            self._pending[seq] = future

        event.wait(timeout=timeout)
        with self._lock:
            self._pending.pop(seq, None)

        return result_holder[0]

    def _send_notification(self, method: str, params: Any = None):
        if not self._process or self._process.poll() is not None:
            return

        msg = {"jsonrpc": "2.0", "method": method, "params": params or {}}
        body = json.dumps(msg)
        header = f"Content-Length: {len(body.encode('utf-8'))}\r\n\r\n"

        try:
            self._process.stdin.write(header + body)
            self._process.stdin.flush()
        except Exception:
            pass

    def did_open(self, file_path: str, language_id: str, content: str):
        if not self._initialized:
            return
        uri = f"file:///{file_path}".replace("\\", "/")
        self._send_notification("textDocument/didOpen", {
            "textDocument": {
                "uri": uri,
                "languageId": language_id,
                "version": 1,
                "text": content,
            }
        })

    def did_change(self, file_path: str, changes: List[Dict[str, Any]], version: int = 1):
        if not self._initialized:
            return
        uri = f"file:///{file_path}".replace("\\", "/")
        self._send_notification("textDocument/didChange", {
            "textDocument": {"uri": uri, "version": version},
            "contentChanges": changes,
        })

    def did_save(self, file_path: str, content: str = ""):
        if not self._initialized:
            return
        uri = f"file:///{file_path}".replace("\\", "/")
        params = {"textDocument": {"uri": uri}}
        if content:
            params["text"] = content
        self._send_notification("textDocument/didSave", params)

    def did_close(self, file_path: str):
        if not self._initialized:
            return
        uri = f"file:///{file_path}".replace("\\", "/")
        self._send_notification("textDocument/didClose", {"textDocument": {"uri": uri}})

    def completion(self, file_path: str, line: int, character: int) -> List[Dict[str, Any]]:
        if not self._initialized:
            return []
        uri = f"file:///{file_path}".replace("\\", "/")
        result = self._send_request_sync("textDocument/completion", {
            "textDocument": {"uri": uri},
            "position": {"line": line, "character": character},
        })
        if isinstance(result, list):
            return result
        if isinstance(result, dict) and "items" in result:
            return result["items"]
        return []

    def hover(self, file_path: str, line: int, character: int) -> Optional[Dict[str, Any]]:
        if not self._initialized:
            return None
        uri = f"file:///{file_path}".replace("\\", "/")
        return self._send_request_sync("textDocument/hover", {
            "textDocument": {"uri": uri},
            "position": {"line": line, "character": character},
        })

    def definition(self, file_path: str, line: int, character: int) -> Any:
        if not self._initialized:
            return None
        uri = f"file:///{file_path}".replace("\\", "/")
        return self._send_request_sync("textDocument/definition", {
            "textDocument": {"uri": uri},
            "position": {"line": line, "character": character},
        })

    def references(self, file_path: str, line: int, character: int) -> List[Dict[str, Any]]:
        if not self._initialized:
            return []
        uri = f"file:///{file_path}".replace("\\", "/")
        return self._send_request_sync("textDocument/references", {
            "textDocument": {"uri": uri},
            "position": {"line": line, "character": character},
            "context": {"includeDeclaration": True},
        }) or []

    def formatting(self, file_path: str) -> List[Dict[str, Any]]:
        if not self._initialized:
            return []
        uri = f"file:///{file_path}".replace("\\", "/")
        return self._send_request_sync("textDocument/formatting", {
            "textDocument": {"uri": uri},
            "options": {"tabSize": 4, "insertSpaces": True},
        }) or []

    def on_diagnostics(self, handler: Callable):
        self._diagnostics_handler = handler

    def stop(self):
        if self._process and self._process.poll() is None:
            self._send_notification("shutdown", {})
            self._send_notification("exit", {})
            try:
                self._process.wait(timeout=5)
            except subprocess.TimeoutExpired:
                self._process.kill()
            self._process = None
            self._initialized = False


class LSPManager:
    def __init__(self):
        self._clients: Dict[str, LSPClient] = {}
        self._workspace_path = ""

    def set_workspace(self, path: str):
        self._workspace_path = path

    def start_python_lsp(self) -> Optional[LSPClient]:
        if "python" in self._clients and self._clients["python"]._initialized:
            return self._clients["python"]

        cmd = self._find_python_lsp()
        if not cmd:
            return None

        client = LSPClient("python", cmd, self._workspace_path)
        if client.start():
            self._clients["python"] = client
            return client
        return None

    def start_dart_lsp(self) -> Optional[LSPClient]:
        if "dart" in self._clients and self._clients["dart"]._initialized:
            return self._clients["dart"]

        dart_path = self._find_dart()
        if not dart_path:
            return None

        cmd = [dart_path, "language-server"]
        client = LSPClient("dart", cmd, self._workspace_path)
        if client.start():
            self._clients["dart"] = client
            return client
        return None

    def start_generic_lsp(self, language_id: str, cmd: List[str]) -> Optional[LSPClient]:
        if language_id in self._clients and self._clients[language_id]._initialized:
            return self._clients[language_id]

        client = LSPClient(language_id, cmd, self._workspace_path)
        if client.start():
            self._clients[language_id] = client
            return client
        return None

    def get_client(self, language_id: str) -> Optional[LSPClient]:
        return self._clients.get(language_id)

    def on_diagnostics(self, language_id: str, handler: Callable):
        client = self._clients.get(language_id)
        if client:
            client.on_diagnostics(handler)

    def stop_all(self):
        for client in self._clients.values():
            client.stop()
        self._clients.clear()

    def _find_python_lsp(self) -> Optional[List[str]]:
        import shutil
        for name in ["pylsp", "pyright-langserver", "pyls", "python-lsp-server"]:
            path = shutil.which(name)
            if path:
                return [path]
        return None

    def _find_dart(self) -> Optional[str]:
        import shutil
        dart_path = shutil.which("dart")
        if dart_path:
            return dart_path
        flutter_path = shutil.which("flutter")
        if flutter_path:
            return flutter_path
        return None


_lsp_manager_instance: Optional[LSPManager] = None


def get_lsp_manager() -> LSPManager:
    global _lsp_manager_instance
    if _lsp_manager_instance is None:
        _lsp_manager_instance = LSPManager()
    return _lsp_manager_instance

File: src/dc/extension-api/lsp/lsp-client.ts
/**
 * Dardcor Code - Language Server Protocol (LSP 3.17) Client Transport (Task 614)
 * Mirrors: vs/workbench/api/common/languageClient (JSON-RPC over stdio)
 */

import * as cp from 'node:child_process';
import { Disposable, IDisposable, toDisposable } from '../../core/lifecycle/disposable';
import { Emitter, Event } from '../../core/events/emitter';

export interface ILspLaunchOptions {
	command: string;
	args?: string[];
	cwd?: string;
	env?: Record<string, string>;
	rootUri?: string;
	workspaceFolders?: string[];
	clientCapabilities?: Record<string, any>;
	initializationOptions?: any;
	trace?: 'off' | 'messages' | 'verbose';
}

export interface ILspInitializeResult {
	capabilities: Record<string, any>;
	serverInfo?: { name: string; version?: string };
}

export interface ILspMessage {
	jsonrpc: '2.0';
	id?: number | string;
	method?: string;
	params?: any;
	result?: any;
	error?: { code: number; message: string; data?: any };
}

export interface ILspRequestHandler {
	(params: any): any | Promise<any>;
}

const JSONRPC = '2.0';

/**
 * LSP 3.17 client. Speaks Content-Length framed JSON-RPC over a child
 * process stdio channel, with a pending-request map for responses,
 * notification handlers, and server-to-client request handlers.
 */
export class LspClient extends Disposable {
	private readonly _onDidNotification = this._register(new Emitter<{ method: string; params: any }>());
	readonly onDidNotification: Event<{ method: string; params: any }> = this._onDidNotification.event;

	private readonly _onDidExit = this._register(new Emitter<{ code: number | null; signal: string | null }>());
	readonly onDidExit: Event<{ code: number | null; signal: string | null }> = this._onDidExit.event;

	private readonly _onDidError = this._register(new Emitter<string>());
	readonly onDidError: Event<string> = this._onDidError.event;

	private _child: cp.ChildProcess | undefined;
	private _seq = 0;
	private _nextHandlerId = 1;
	private readonly _pending = new Map<number, { resolve: (value: any) => void; reject: (err: Error) => void; method: string }>();
	private readonly _notificationHandlers = new Map<string, Set<(params: any) => void>>();
	private readonly _requestHandlers = new Map<string, ILspRequestHandler>();
	private _buffer = '';
	private _initializeResult: ILspInitializeResult | undefined;

	public get isRunning(): boolean {
		return !!this._child && !this._child.killed;
	}

	public get initializeResult(): ILspInitializeResult | undefined {
		return this._initializeResult;
	}

	public async start(options: ILspLaunchOptions): Promise<ILspInitializeResult> {
		if (this._child) {
			throw new Error('LSP client sudah berjalan');
		}
		const child = cp.spawn(options.command, options.args ?? [], {
			cwd: options.cwd,
			env: options.env ? { ...process.env, ...options.env } : process.env,
			stdio: ['pipe', 'pipe', 'pipe'],
			windowsHide: true
		});
		this._child = child;
		child.stdout?.setEncoding('utf8');
		child.stdout?.on('data', (chunk: string) => this._processChunk(chunk));
		child.stderr?.setEncoding('utf8');
		child.stderr?.on('data', (chunk: string) => this._onDidError.fire(chunk));
		child.on('error', (err: Error) => {
			this._onDidError.fire(`Server LSP gagal dijalankan: ${err.message}`);
			this._rejectAll(new Error(err.message));
		});
		child.on('close', (code, signal) => {
			this._child = undefined;
			this._rejectAll(new Error('Server LSP berhenti'));
			this._onDidExit.fire({ code, signal });
		});

		const result = await this.initialize(options);
		this._initializeResult = result;
		this.notify('initialized', {});
		return result;
	}

	public async initialize(options: ILspLaunchOptions): Promise<ILspInitializeResult> {
		const result = await this.request<ILspInitializeResult>('initialize', {
			processId: process.pid,
			rootUri: options.rootUri ?? null,
			workspaceFolders: (options.workspaceFolders ?? []).map(f => ({ uri: f, name: f.split('/').pop() ?? f })),
			capabilities: options.clientCapabilities ?? LspClient.defaultClientCapabilities(),
			initializationOptions: options.initializationOptions,
			trace: options.trace ?? 'off'
		});
		return { capabilities: result.capabilities ?? {}, serverInfo: result.serverInfo };
	}

	public request<T = any>(method: string, params?: any): Promise<T> {
		return new Promise<T>((resolve, reject) => {
			if (!this.isRunning) {
				reject(new Error('Server LSP tidak berjalan'));
				return;
			}
			const id = this._seq++;
			this._pending.set(id, { resolve, reject, method });
			this._send({ jsonrpc: JSONRPC, id, method, params });
		});
	}

	public notify(method: string, params?: any): void {
		if (!this.isRunning) {
			return;
		}
		this._send({ jsonrpc: JSONRPC, method, params });
	}

	public onNotification(method: string, handler: (params: any) => void): IDisposable {
		let set = this._notificationHandlers.get(method);
		if (!set) {
			set = new Set();
			this._notificationHandlers.set(method, set);
		}
		set.add(handler);
		return toDisposable(() => set?.delete(handler));
	}

	public onRequest(method: string, handler: ILspRequestHandler): IDisposable {
		this._requestHandlers.set(method, handler);
		return toDisposable(() => this._requestHandlers.delete(method));
	}

	public async shutdown(): Promise<void> {
		if (!this.isRunning) {
			return;
		}
		try {
			await this.request('shutdown', null);
			this.notify('exit', null);
		} catch {
			// server sudah berhenti
		}
	}

	public stop(): void {
		this._child?.kill();
		this._child = undefined;
		this._rejectAll(new Error('LSP client dihentikan'));
	}

	public didOpen(uri: string, languageId: string, version: number, text: string): void {
		this.notify('textDocument/didOpen', {
			textDocument: { uri, languageId, version, text }
		});
	}

	public didChange(uri: string, version: number, changes: Array<{ range?: any; rangeLength?: number; text: string }>): void {
		this.notify('textDocument/didChange', {
			textDocument: { uri, version },
			contentChanges: changes
		});
	}

	public didClose(uri: string): void {
		this.notify('textDocument/didClose', { textDocument: { uri } });
	}

	public didSave(uri: string, text?: string): void {
		this.notify('textDocument/didSave', {
			textDocument: { uri },
			text
		});
	}

	public static defaultClientCapabilities(): Record<string, any> {
		return {
			workspace: {
				workspaceFolders: true,
				configuration: true,
				applyEdit: true,
				didChangeConfiguration: { dynamicRegistration: true },
				didChangeWatchedFiles: { dynamicRegistration: true },
				symbol: { dynamicRegistration: true }
			},
			textDocument: {
				publishDiagnostics: { relatedInformation: true, codeDescription: true, dataSupport: true },
				synchronization: { dynamicRegistration: true, willSave: true, willSaveWaitUntil: true, didSave: true },
				completion: { completionItem: { snippetSupport: true, documentationFormat: ['markdown', 'plaintext'], insertReplaceSupport: true } },
				hover: { contentFormat: ['markdown', 'plaintext'] },
				definition: { linkSupport: true },
				references: {},
				rename: { prepareSupport: true },
				formatting: { dynamicRegistration: true },
				codeAction: { codeActionLiteralSupport: { codeActionKind: { valueSet: ['quickfix', 'refactor', 'source'] } } },
				codeLens: { dynamicRegistration: true },
				inlayHint: { dynamicRegistration: true },
				semanticTokens: { dynamicRegistration: true, tokenTypes: [], tokenModifiers: [] }
			},
			window: { workDoneProgress: true }
		};
	}

	private _send(message: ILspMessage): void {
		const child = this._child;
		if (!child?.stdin) {
			throw new Error('Server LSP tidak berjalan');
		}
		const body = JSON.stringify(message);
		child.stdin.write(`Content-Length: ${Buffer.byteLength(body, 'utf8')}\r\n\r\n${body}`);
	}

	private _processChunk(chunk: string): void {
		this._buffer += chunk;
		while (true) {
			const headerEnd = this._buffer.indexOf('\r\n\r\n');
			if (headerEnd === -1) {
				if (this._buffer.length > 65536) {
					this._buffer = this._buffer.substring(this._buffer.length - 4096);
				}
				return;
			}
			const headerText = this._buffer.substring(0, headerEnd);
			const lengthMatch = /Content-Length:\s*(\d+)/i.exec(headerText);
			if (!lengthMatch) {
				this._buffer = this._buffer.substring(headerEnd + 4);
				continue;
			}
			const contentLength = parseInt(lengthMatch[1], 10);
			if (this._buffer.length < headerEnd + 4 + contentLength) {
				return;
			}
			const body = this._buffer.substring(headerEnd + 4, headerEnd + 4 + contentLength);
			this._buffer = this._buffer.substring(headerEnd + 4 + contentLength);
			this._handleMessage(body);
		}
	}

	private _handleMessage(body: string): void {
		let message: ILspMessage;
		try {
			message = JSON.parse(body);
		} catch {
			this._onDidError.fire('Pesan LSP tidak valid');
			return;
		}
		if (message.id !== undefined && message.method === undefined) {
			const pending = this._pending.get(message.id as number);
			if (pending) {
				this._pending.delete(message.id as number);
				if (message.error) {
					pending.reject(new Error(`LSP ${pending.method}: ${message.error.message}`));
				} else {
					pending.resolve(message.result);
				}
			}
		} else if (message.id !== undefined && message.method) {
			this._handleServerRequest(message);
		} else if (message.method) {
			const handlers = this._notificationHandlers.get(message.method);
			if (handlers) {
				for (const handler of handlers) {
					try {
						handler(message.params);
					} catch (err) {
						console.error(`[lsp-client] Handler '${message.method}' gagal:`, err);
					}
				}
			}
			this._onDidNotification.fire({ method: message.method, params: message.params });
		}
	}

	private _handleServerRequest(message: ILspMessage): void {
		const handler = this._requestHandlers.get(message.method ?? '');
		if (!handler) {
			this._send({ jsonrpc: JSONRPC, id: message.id, error: { code: -32601, message: `Method not found: ${message.method}` } });
			return;
		}
		try {
			const result = handler(message.params);
			if (result && typeof result.then === 'function') {
				result.then(
					(value: any) => this._send({ jsonrpc: JSONRPC, id: message.id, result: value }),
					(err: any) => this._send({ jsonrpc: JSONRPC, id: message.id, error: { code: -32603, message: String(err?.message ?? err) } })
				);
			} else {
				this._send({ jsonrpc: JSONRPC, id: message.id, result });
			}
		} catch (err) {
			this._send({ jsonrpc: JSONRPC, id: message.id, error: { code: -32603, message: String(err instanceof Error ? err.message : err) } });
		}
	}

	private _rejectAll(err: Error): void {
		for (const [, pending] of this._pending) {
			pending.reject(err);
		}
		this._pending.clear();
	}
}
